Meetup API Client: Use OAuth for authentication instead of an API key

coreymckrill opened the pull request:
Meetup.com [announced](https://groups.google.com/forum/#!topic/meetup-api/R__7mPzWJc0) that it's API key method of authentication would be going away on August 15th, 2019. That's what we've been using, so we need to switch to their OAuth2 method instead.

In order to avoid having to manually redirect to Meetup.com and login with credentials before each time we need to pull the data from the API, we are using the [Server Flow with User Credentials](https://www.meetup.com/meetup_api/auth/#oauth2servercredentials) approach. Unfortunately this requires a non-standard grant type, and a strange extra step of using an access token to request a separate, more special access token. As such, existing OAuth client libraries don't support it out of the box (that I'm aware of). Instead, this PR adds a very simple, custom OAuth2 client.

coreymckrill committed:
Refactor Meetup_Client to extend the API_Client base class

coreymckrill committed:
Tweak client debug mode

coreymckrill committed:
API Client: Make the cli_command method static

This function doesn't require any state from a class instance so
it should be static.

coreymckrill committed:
API Client: Sync layout of header statements with other API client files

coreymckrill committed:
Add new `Meetup_OAuth2_Client` class

coreymckrill committed:
Meetup Client: Make the throttle method static

The throttle method doesn't require any state from a class instance
so it's ok to be static.

coreymckrill committed:
Meetup Client: Improve formatting of error handler method

coreymckrill committed:
Meetup Client: Switch cli_message calls to static

coreymckrill committed:
Meetup Client: Update to use new OAuth client instead of API key

coreymckrill committed:
Docker: Add new placeholder constants for Meetup creds

iandunn commented on public_html/wp-content/mu-plugins/utilities/class-meetup-oauth2-client.php:
```diff
@@ -0,0 +1,460 @@
+			default:
+				$request[ $type ] = $code;
+				break;
+		}
```
`SHOULD` - It's not obvious to me what's going on here. Is one of these for getting a new code, and the other for renewing an expired code? It'd be helpful to add inline comments for each saying what they're used for.

coreymckrill replied:
Fixed in 0f24fca, but then I ended up doing a significant refactor in ea24b48. Might be worth taking another look to see if it's understandable now.

iandunn replied:
Yeah, this looks good. Kudos for persevering through their inaccurate docs!

iandunn reviewed (approved):
LGTM, good job digging into their strange implementation on short notice so this critical part of our program can continue working!

coreymckrill committed:
OAuth client: Clarify parameters for access token request payload

coreymckrill committed:
OAuth client: Avoid errors and false positives in `is_valid_token`

This eliminates scenarios where `$required_properties` was not defined
before usage, and where using an invalid token `$type` could return
a truthy value.

coreymckrill committed:
OAuth client: Refactor to correctly use refresh_token flow

The previous refresh_token flow was useless because the documentation
for it is inaccurate. Blerg.

coreymckrill committed:
OAuth client: Remove unnecessary condition

coreymckrill committed:
OAuth client: Check expiration of locally cached token before using

coreymckrill committed:
Add documentation about multiple locations of files

coreymckrill commented:
Merged via aa9e9247f19036dddc64992082114cf23d4d3f63
